src/components/Admin/Trainer/ControllerDataset.js

```diff
@@ -25,51 +22,82 @@ export class ControllerDataset {
     return itemObj;
   }
 
-  async addExamples(examples) {
-    if (examples.length < 1) {
-      this.setReadyStatus('Please provide at least one image.');
-      return;
-    }
+  getCollectionReference = async name => {
+    return await this.db.collection(name);
+  };
+
+  getItemReference = async label => {
+    return await this.db.collection('item_data').doc(label);
+  };
+
+  getItemCount = async itemReference => {
+    return await itemReference
+      .get()
+      .then(snapshot => snapshot.data().count)
+      .catch(() => 0);
+  };
+
+  setItemCount = async (itemReference, count) => {
+    await itemReference.set({
+      count
+    });
+  };
+
+  changeItemCount = (label, delta) => {
+    this.getItemReference(label).then(item =>
+      this.getItemCount(item).then(count =>
+        this.setItemCount(item, count + delta)
+      )
+    );
+  };
+
+  deleteImage = async dataset => {
+    await this.db
+      .collection('training_data')
+      .doc(dataset.id)
+      .delete();
+
+    await this.changeItemCount(dataset.item, -1);
+  };
 
-    const item = await this.db
-      .collection('item_data')
-      .doc(examples[0].label)
-      .get();
+  trustImage = async dataset => {
+    await this.db
+      .collection('training_data')
+      .doc(dataset.id)
+      .update({trusted: true});
 
+    await this.changeItemCount(dataset.item, 1);
+  };
+
+  addImage = (image, trusted) => {
     this.db
-      .collection('item_data')
-      .doc(examples[0].label)
-      .set({
-        count: item.exists
-          ? item.data().count + examples.length
-          : examples.length
+      .collection('training_data')
+      .add({
+        img: image.img,
+        activation: image.activation.dataSync().join(','),
+        label: image.label,
+        random: Math.random(),
+        timestamp: Date.now(),
+        trusted
+      })
+      .then(() => {
+        image.activation.dispose();
```

Comment:
We could use `Promise.prototype.finally` here, unfortunately safari doesn't support this though and it's probably not worth introducing a polyfill for it.

It'd look like

```
.add({
 // ...
}).finally(() => {
  image.activation.dispose();
})
```